### DISCUSSION TITLE
How to compare train vs test performance when using make_reduction?

### DISCUSSION THREAD

**Discussion by ericjb** [alan-turing-institute/sktime#7357] [Q&A] [answered]
A standard test for overfitting is to compare performance - e.g. MAPE, MAE, RMSE - between the training and test periods. How does one get these metrics for the training period when using make_reduction?

Here is a specific example

```python
from sktime.forecasting.model_selection import temporal_train_test_split
from sklearn.neural_network import MLPRegressor
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.compose import EnsembleForecaster
from sktime.performance_metrics.forecasting import mean_absolute_percentage_error

from sktime.datasets import load_airline
y = load_airline()

TEST_SIZE = 24
train, test = temporal_train_test_split(y, test_size=TEST_SIZE)
N_NETWORKS = 20
regressor = MLPRegressor(hidden_layer_sizes=(7,), shuffle=False, activation='relu', max_iter=3000, solver='lbfgs')
mlp_fc = make_reduction(regressor, window_length=12, strategy="recursive")

ensemble = EnsembleForecaster([(f"model_{i}", mlp_fc) for i in range(N_NETWORKS)])
ensemble.fit(train)

pred = ensemble.predict(test.index)
print(f'Test Period: MAPE  = {mean_absolute_percentage_error(test, pred)*100:.2f}%')
print(f'Train Period: MAPE  = ???')
```

**Comment by ericjb** [comment]
The following calculates MAPE for the training and test periods using mainly sklearn functionality (with only SlidingWindowSplitter from sktime). It is not clear to me why this cannot be done using (almost) only sktime functionality. I would think this is a fairly common use case. Thanks for any explanation - and hopefully pointers how to accomplish this in a more sktime-centric manner.

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.ensemble import VotingRegressor
from sktime.forecasting.model_selection import SlidingWindowSplitter
from sktime.datasets import load_airline
from sktime.utils import plot_series

import matplotlib
matplotlib.use('TkAgg')

# Load the dataset
z = load_airline()

# Set the random seed for reproducibility
np.random.seed(45)

# Create the SlidingWindowSplitter
splitter = SlidingWindowSplitter(fh=[1], window_length=12, step_length=1, start_with_window=True)

# Generate the indices for the sliding windows
indices = list(splitter.split(z))

# Extract the feature matrix X and target vector y using the indices
X = np.array([z.iloc[train].values for train, _ in indices])
y = np.array([z.iloc[test].values[0] for _, test in indices])

# Define the split point
TEST_SIZE = 24
SPLIT_POINT = len(X) - TEST_SIZE

# Manually split the data into training and testing sets
X_train, X_test = X[:SPLIT_POINT], X[SPLIT_POINT:]
y_train, y_test = y[:SPLIT_POINT], y[SPLIT_POINT:]

# Create an ensemble of 20 MLPRegressors using VotingRegressor
N_NETWORKS = 20
estimators = [(f"mlp_{i}", MLPRegressor(hidden_layer_sizes=(7,), shuffle=False, 
                activation='relu', max_iter=3000, solver='lbfgs')) for i in range(N_NETWORKS)]
ensemble = VotingRegressor(estimators)

# Fit the ensemble on the training data
ensemble.fit(X_train, y_train)

# Predict on the training set
y_train_pred = ensemble.predict(X_train)

# Rolling forward predictions for the test set
y_test_pred = []
current_window = X_train[-1].tolist()

for i in range(TEST_SIZE):
    next_pred = ensemble.predict([current_window])[0]
    y_test_pred.append(next_pred)
    current_window = current_window[1:] + [next_pred]

y_series = pd.Series(y, index=z.index[12:])
y_pred_series = pd.Series(np.array(y_test_pred), index=z.index[-TEST_SIZE:])

# Compute MAPE for both training and testing sets
mape_train = mean_absolute_percentage_error(y_train, y_train_pred)
mape_test = mean_absolute_percentage_error(y_test, y_test_pred)

print(f'Train Period: MAPE = {mape_train * 100:.2f}%')
print(f'Test Period: MAPE = {mape_test * 100:.2f}%')

plot_series(y_series, y_pred_series, labels=["y", "y_pred"])
plt.show()

```

**Reply by ericjb** [reply]
I modified the code in this comment to correctly calculate the out-of-sample forecasts using sliding windows that use forecasted values. This is consistent with how sktime does such forecasts.

**Comment by ericjb** [comment]
I discovered the following:
1. predict is explicitly disabled for training period horizons for reduced forecasters (i.e. those obtained via make_reduction). This is done in forecasting/compose/_reduce.py where the class _Reducer sets the tag "capability:insample" to False.
2. In the same file (_reduce.py) the method _predict_in_sample is a stub (it has the one line 'pass')
3. The _Reducer class is a derived class of _BaseWindowForecaster (defined in forecasting/base/_sktime.py). The _BaseWindowForecaster class also defines the method _predict_in_sample and it seems to be correct for _Reduce objects.
4. I made the following two "small" modifications and I am able to get predict values for the training period.
    a. set capability:insample to True
    b. modify _predict_in_sample in _Reducer, replacing 'pass' by
```python
          cutoff = self._cutoff
          return super()._predict_in_sample(fh.to_in_sample(cutoff), **kwargs)
```
Would a Pull Request with these two changes be accepted? I assume not, based on some problematic cases of which I have no understanding. But after studying the code it seems that in many cases these changes would actually be fine.  Can we isolate the problematic cases so that capability:insample is effectively False for them, but is True for the non-problematic cases? That would allow for the method _predict_in_sample - which is already implemented! - to be used. It's a shame that it is blocked. Given some guidance I would be happy to work on this. Perhaps there are some tests that failed that led to the capability:insample being set to False.

**Comment by fkiraly** [comment]
Sorry for the late reply, I had longer offline periods last week.

There are multiple things to comment on:

* For general forecasters, the programmatic way to get in-sample metrics is to simply call `predict` with in-sample indices. However, the first generation of reducers in `sktime` do not support in-sample forecasts still. You can use the second-generation rework `DirectReductionForecaster` though, which does support in-sample forecasts.
* you seem to have found out how to enable in-sample forecasts for the first generation reducer! That is great! We have been trying to unpack the somewhat obscure hierarchy of inheritance for a while, this turned out very hard to understand as the code has been modified by many people in an organic way, without clear documentation or comments...
    * so, indeed the PR would be accepted and much appreciated! Kudos for finding out how to do this!
    * what would also be appreciated if you added comments or enhance docstrings so people after you would also understand what to do, or which parent or child the called methods are actually coming from when called.
    * more generally, do you see a way to "flatten" the inheritance structure for the reducers, and make the code and code structure clearer in the first generation reducer? (everying accessible via `make_reduction`)

Always great to read your thoughts, @ericjb!

Also, in case you want to understand the larger complex of feature requests and issues related to the reduction forecasters, we have collected them in a project board: https://github.com/orgs/sktime/projects/5

**Reply by ericjb** [reply]
Hi @fkiraly - welcome back! On further testing there is something weird going on with my change to the "first-generation" reducer to get in-sample, and I have not yet completely figured it out.  Regarding your comments:
1. If there is a second-generation, which presumably is an improvement, why should I (or anyone else) invest time in the first-generation?
2. I started investigating the DirectReductionForecaster and I am a bit confused. I tried to create an EnsembleForecaster of a forecaster created using DirectReductionForecaster. The tags for EnsembleForecaster include "requires-fh-in-fit": False, yet when I try to fit the EnsembleForecaster it complains that fh must be specified. 
3. Also, in make_reduction, there is a parameter 'strategy' which can be set to "recursive". Is recursive done with the DirectReductionForecaster? Can you explain, or point to an example, where forecasts are made for the training and test periods?
Thanks

**Reply by ericjb** [reply]
Update: I investigated further and I found that DirectReductionForecaster sets the tag "requires-fh-in-fit" to True. Why? I guess it has to do with creating the tabular inputs to the sklearn. But then it gets back to point 3 above,i.e. if you use fh=1 (relative) for the fitting, then you need to do recursive forecasts for multiple forecasts (for the test period, as opposed to the training period), and I have not yet found where that is provided for in predict. Unless I am missing something.

**Reply by ericjb** [reply, marked as answer]
[tmpprofile.txt](https://github.com/user-attachments/files/17692018/tmpprofile.txt)
**Additional Update:** I decided to try **RecursiveReductionForecaster** because that inherits from _ReducerMixin and not _Reducer. The latter sets **capability:insample** to False. 
The good news is that this class produced both insample and out-of-sample results! The bad news is that **the out-of-sample results took a really long time**. Here is the timing as produced by some print statements in the program. Note that the test set has size 36,  and the training set size 124; the forecaster is an ensemble of 20 mlp sklearn forecasters (each with one hidden layer with seven nodes; window size 12).

```
2024-11-10 17:04:04 - here 1
2024-11-10 17:04:04 - here 2 - before fit
2024-11-10 17:04:06 - here 3 - after fit before out-of-sample
2024-11-10 17:05:02 - here 4 - after out-of-sample before predict insample
2024-11-10 17:05:04 - here 5 - after in-sample
```

This shows: a) the fit took 2 seconds, b) the 36 out-of-sample forecasts took 56 seconds; and c) the 112 in-sample forecasts took 2 seconds. **This makes no sense. The out-of-sample should be well under 2 seconds.**

I then ran a **profiler** on the forecasting (including both in-sample and out-of-sample). Here is an excerpt of the output (which has longer timings because it includes the extra time taken by the profiler - which approximately doubled the run time) [this output is also attached]

```
2024-11-10 15:53:03 - here 1
2024-11-10 15:53:03 - here 2 - before fit
2024-11-10 15:53:05 - here 3 - after fit before out-of-sample
2024-11-10 15:55:08 - here 4 - after out-of-sample before predict insample
2024-11-10 15:55:13 - here 5 - after in-sample

263633590 function calls (233038527 primitive calls) in 127.208 seconds

Ordered by: internal time

           ncalls  tottime   cumtime    filename:lineno(function)
  28775078/651190   21.263    36.898    /usr/lib/python3.10/copy.py:128(deepcopy)
          3255320   13.893    13.899    ./pandas/core/arrays/datetimes.py:2712(_generate_range)
           601330    5.498    35.187    /usr/lib/python3.10/copy.py:227(_deepcopy_dict)
         58041307    4.233     4.233    {method 'get' of 'dict' objects}
23202933/22971461    3.385     5.195    {built-in method builtins.isinstance}
    324000/319618    2.988    21.623    ./pandas/core/series.py:389(__init__)
         31749598    2.346     2.346    {built-in method builtins.id}
         27612804    1.862     1.862    /usr/lib/python3.10/copy.py:182(_deepcopy_atomic)
           384186    1.691     5.384    ./pandas/core/construction.py:517(sanitize_array)
           296224    1.338    23.225    ./skbase/base/_tagmanager.py:16(_get_class_flags)
  6736907/4880672    1.251     1.837    {built-in method builtins.len}
  7198226/7092726    1.227     1.450    {built-in method builtins.getattr}
           409728    1.222     3.597    ./pandas/core/dtypes/common.py:1596(pandas_dtype)
           653756    0.918     2.830    ./pandas/core/generic.py:6301(__setattr__)
           529208    0.910     1.198    ./pandas/core/generic.py:278(__init__)
            22840    0.883    14.782    ./pandas/core/arrays/datetimes.py:468(<listcomp>)
  3431360/3429200    0.872     1.314    {built-in method builtins.hasattr}
           279728    0.840     0.840    {built-in method numpy.array}
           266188    0.832     1.286    ./pandas/core/generic.py:6236(__finalize__)
           157444    0.825     0.825    {method 'reduce' of 'numpy.ufunc' objects}
          1162274    0.799     1.019    /usr/lib/python3.10/copy.py:243(_keep_alive)
            22860    0.730    16.515    ./pandas/core/indexes/datetimelike.py:497(shift)
          2063985    0.692     1.637    ./pandas/core/dtypes/generic.py:42(_instancecheck)
           560944    0.684     2.005    /usr/lib/python3.10/copy.py:201(_deepcopy_list)
          2063985    0.660     0.893    ./pandas/core/dtypes/generic.py:37(_check)
            63786    0.641     9.903    ./sktime/datatypes/_check.py:131(check_is_mtype)
           319616    0.621     2.459    ./pandas/core/internals/managers.py:1863(from_array)
          1602746    0.597     0.597    {method 'update' of 'dict' objects}
           243420    0.554     1.108    ./pandas/core/arrays/datetimes.py:545(_box_func)
          1834204    0.545     0.783    ./pandas/core/indexes/base.py:909(__len__)
           232491    0.535     0.699    ./pandas/core/indexes/base.py:649(_simple_new)
           247952    0.517    19.484    ./pandas/core/generic.py:6432(dtypes)
            51006    0.509     0.836    ./pandas/core/indexes/base.py:842(_engine)
           273988    0.479     0.525    ./pandas/core/indexes/base.py:3777(get_loc)
           231664    0.478    33.144    ./skbase/base/_tagmanager.py:88(_get_flags)
           247954    0.464     1.627    ./pandas/core/internals/managers.py:287(get_dtypes)
           338720    0.458     1.630    ./pandas/core/generic.py:807(_set_axis)
           366146    0.452     0.832    ./pandas/core/internals/blocks.py:2645(maybe_coerce_values)
           364600    0.431     0.833    ./pandas/core/internals/blocks.py:2716(new_block)
           212762    0.424     1.747    ./pandas/core/series.py:1095(__getitem__)
           324768    0.422     0.879    ./pandas/_config/config.py:127(_get_single_key)
    285486/203504    0.416     0.953    ./sktime/datatypes/_registry.py:112(mtype_to_scitype)
            43910    0.414     2.733    ./pandas/core/indexes/base.py:475(__new__)
           255255    0.405     0.670    /usr/lib/python3.10/warnings.py:181(_add_filter)
           232764    0.398     0.398    ./sktime/datatypes/_registry.py:156(<listcomp>)
           361451    0.398     1.953    ./pandas/core/indexes/base.py:5373(__getitem__)
           100084    0.397     2.474    ./pandas/core/internals/managers.py:317(apply)

            RMSE        MAE      MAPE      MASE
Train  76.876218  60.768292  3.521961  0.501667
          RMSE        MAE      MAPE      MASE
Test  96.05723  78.255368  3.897771  0.624556
```

I will be investigating to see how to reduce the time, and would be happy to receive hints on what might be causing the issue and/or how to fix/improve it.

**Reply by ericjb** [reply]
I found a solution to the slowness. In this particular example, the speedup is 50x: in fact, at the 1-second resolution of the time stamps, the out-of-sample forecasts took no time at all! (versus 56 seconds). The fix involved copying and modifying the _predict_out_of_sample method from a different class. It worked. :-)

I will update issue #3224 which had already noticed and commented on the slowness of this class.

```
2024-11-11 19:17:02 - here 1
2024-11-11 19:17:02 - here 2 - before fit
2024-11-11 19:17:06 - here 3 - after fit before out-of-sample
2024-11-11 19:17:06 - here 4 - after out-of-sample before predict insample
2024-11-11 19:17:09 - here 5 - after in-sample
            RMSE        MAE      MAPE      MASE
Train  76.876218  60.768292  3.521961  0.501667
          RMSE        MAE      MAPE      MASE
Test  96.05723  78.255368  3.897771  0.624556
```
